### PR TITLE
No more ifup/ifdown on Ubuntu >=17.10

### LINKED ISSUE #962: Add support for Ubuntu 1710
1710 does not have ifup/ifdown commands. We need to use netplan to bounce the interface.

### PR DESCRIPTION
**Pull request by boumenot**
Closes #962

### CODE REVIEW

**Comment by boumenot** [comment]
I am validating this change in DCR.  Based on my reading this is the correct way to bounce the interface, and force a renew of DHCP.

File: azurelinuxagent/common/osutil/ubuntu.py [hunk 1]
```diff
@@ -60,6 +70,30 @@ def register_agent_service(self):
     def unregister_agent_service(self):
         return shellutil.run("systemctl mask walinuxagent", chk_err=False)
 
+
+class UbuntuOSUtil(Ubuntu16OSUtil):
+    def __init__(self):
+        super(UbuntuOSUtil, self).__init__()
+
+    def restart_if(self, ifname, retries=3, wait=5):
```
Comment (hglkrijger) [inline azurelinuxagent/common/osutil/ubuntu.py]: I am hesitant to change this for the 16x images as well. Can we limit to 1710?

File: azurelinuxagent/common/osutil/ubuntu.py [hunk 2]
```diff
@@ -60,6 +70,30 @@ def register_agent_service(self):
...
+        """
+        Restart an interface by bouncing the link. systemd-networkd observes
+        this event, and forces a renew of DHCP.
+        """
+        retry_limit=retries+1
+        for attempt in range(1, retry_limit):
+            return_code=shellutil.run("ip link set {0} down && ip link set {0} down".format(ifname))
```
Comment (hglkrijger) [inline azurelinuxagent/common/osutil/ubuntu.py]: && link up

File: azurelinuxagent/common/osutil/ubuntu.py [hunk 3]
```diff
@@ -60,6 +70,30 @@ def register_agent_service(self):
...
+            if return_code == 0:
+                return
+            logger.warn("failed to restart {0}: return code {1}".format(ifname, return_code))
```
Comment (hglkrijger) [inline azurelinuxagent/common/osutil/ubuntu.py]: Please add dcr tests before merging

File: azurelinuxagent/common/osutil/ubuntu.py [hunk 4]
```diff
@@ -50,7 +56,11 @@ def get_dhcp_pid(self):
         ret = shellutil.run_get_output("pidof dhclient3", chk_err=False)
         return ret[1] if ret[0] == 0 else None
 
-class UbuntuOSUtil(Ubuntu14OSUtil):
+
+class Ubuntu16OSUtil(Ubuntu14OSUtil):
```
Comment (hglkrijger) [inline azurelinuxagent/common/osutil/ubuntu.py]: This will be either 17 or 18, I think. Convention has been to use LTS versions but I am ok either way.

Comment (boumenot) [reply azurelinuxagent/common/osutil/ubuntu.py]: I took the approach that the latest version of Ubuntu is current, and if you need to make a change you use the version number.

 * UbuntuOSUtil is for >= 17.10
 * Ubuntu16OSUtil was created an now represents previous versions.  

This seems to be the style of the current code.

File: azurelinuxagent/common/osutil/ubuntu.py (hunk 1 above)
Comment (boumenot) [reply azurelinuxagent/common/osutil/ubuntu.py]: This change is only for 17.10 and above.

File: azurelinuxagent/common/osutil/ubuntu.py (hunk 3 above)
Comment (boumenot) [reply azurelinuxagent/common/osutil/ubuntu.py]: What is not already covered, that needs to be covered by this change?

**Pull request merged**